File: 2020/java/Advent02.java
import java.util.*;
import java.util.regex.Pattern;
import java.util.stream.*;

public class Advent02 extends Advent {

  public static final String REGEX = "(\\d+)-(\\d+) (.): (.+)";

  public Advent02() {
    super(2);
  }

  @Override
  protected void parseInput() {

  }

  @Override
  protected Object part1() {
    return "" + input.stream()
      .filter(this::isValid)
      .count();
  }

  boolean isValid(String s) {
    var p = Pattern.compile(REGEX);
    var m = p.matcher(s);
    m.find();
    var sCount = m.group(4).chars().filter(c -> c == m.group(3).charAt(0)).count();
    int min = Integer.parseInt(m.group(1));
    int max = Integer.parseInt(m.group(2));
    return sCount >= min && sCount <= max;
  }

  @Override
  protected Object part2() {
    return "" + input.stream()
      .filter(this::isValid2)
      .count();
  }

  boolean isValid2(String s) {
    try {
      var p = Pattern.compile(REGEX);
      var m = p.matcher(s);
      m.find();
      int min = Integer.parseInt(m.group(1));
      int max = Integer.parseInt(m.group(2));
      var pw = m.group(4);
      char c = m.group(3).charAt(0);
      return pw.charAt(min-1) == c ^ pw.charAt(max-1) == c;
    } catch (Exception e) {
      return false;
    }
  }
}

File: 2020/python/advent_02.py
from common.advent_lib import *
import itertools as iter
import re

lines = lines(2)

def matches(s):
    min, max, char, pw = re.match(r"(\d+)-(\d+) (.): (.+)", s).groups()
    count = len(list(re.finditer(re.compile(char), pw)))
    return count >= int(min) and count <= int(max)


print("Part 1: ", len(list(filter(matches, lines))))

def matches2(s):
    min, max, char, pw = re.match(r"(\d+)-(\d+) (.): (.+)", s).groups()
    return (pw[int(min)-1] == char) ^ (pw[int(max)-1] == char)

print("Part 2: ", len(list(filter(matches2, lines))))
